From d0659284675dc8121ec0dba6e9477bf0f7427b0d Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Thu, 16 Jul 2020 11:59:41 +0800
Subject: [PATCH] PONRTSYS-4557:GSWAPI: support for configure TPID for VLAN
 tagged packets

4 TPID mode must be enabled on VLANTPID repective DMA port
one of the 4 values 0x8100 and other 3 programmable via
FDMA_VTETYPE, FDMA_VTETYPE2 and FDMA_VTETYPE3.
VLAN TPID for treatment can configured per rule to be one of the 4 TPID's
Treatment part enabled when Extended VLAN tagging pattern match or triggered by flow
classification, Extended VLAN table serched and action in treatment part is taken
then new outer and inner VLAN tag is added to the frame.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 86 ++++++++++++++++++++++
 1 file changed, 86 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 65c3bbdf8a01..088c0d7e2269 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -19818,6 +19818,12 @@ GSW_return_t GSW_ExtendedVlanSet(void *cdev, GSW_EXTENDEDVLAN_config_t *parm)
 
 	/** Extended VLAN Table Filter **/
 	/*Program Outer VLAN Filter*/
+
+	/*If TPID mode is 4 and Meter treatment is disabled*/
+	if (!parm->sTreatment.bNewMeterEnable &&
+	    parm->sFilter.bOriginalPacketFilterMode)
+		tbl_prog.val[5] |= (1 << 0x4);
+
 	switch (parm->sFilter.sOuterVlan.eType) {
 	/** There is tag and criteria applies. */
 	case GSW_EXTENDEDVLAN_FILTER_TYPE_NORMAL:
@@ -19870,6 +19876,26 @@ GSW_return_t GSW_ExtendedVlanSet(void *cdev, GSW_EXTENDEDVLAN_config_t *parm)
 
 		/** TPID is global configured value. */
 		case GSW_EXTENDEDVLAN_FILTER_TPID_VTETYPE:
+			if (!parm->sTreatment.bNewMeterEnable) {
+				switch (parm->sFilter.eFilter_4_Tpid_Mode) {
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_1:
+					tbl_prog.val[5] |= (0 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_2:
+					tbl_prog.val[5] |= (1 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_3:
+					tbl_prog.val[5] |= (2 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_4:
+					tbl_prog.val[5] |= (3 << 0x0);
+					break;
+				}
+			}
+
 			switch (parm->sFilter.sOuterVlan.eDei) {
 			/** Do not filter (don't care) key 0 BIT 14:12 set 5 */
 			case GSW_EXTENDEDVLAN_FILTER_DEI_NO_FILTER:
@@ -19995,6 +20021,26 @@ GSW_return_t GSW_ExtendedVlanSet(void *cdev, GSW_EXTENDEDVLAN_config_t *parm)
 
 		/** TPID is global configured value. */
 		case GSW_EXTENDEDVLAN_FILTER_TPID_VTETYPE:
+			if (!parm->sTreatment.bNewMeterEnable) {
+				switch (parm->sFilter.eFilter_4_Tpid_Mode) {
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_1:
+					tbl_prog.val[5] |= (0 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_2:
+					tbl_prog.val[5] |= (1 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_3:
+					tbl_prog.val[5] |= (2 << 0x0);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_4:
+					tbl_prog.val[5] |= (3 << 0x0);
+					break;
+				}
+			}
+
 			switch (parm->sFilter.sInnerVlan.eDei) {
 			/** Do not filter (don't care) key 2 BIT 14:12 set 5 */
 			case GSW_EXTENDEDVLAN_FILTER_DEI_NO_FILTER:
@@ -20204,6 +20250,26 @@ GSW_return_t GSW_ExtendedVlanSet(void *cdev, GSW_EXTENDEDVLAN_config_t *parm)
 
 		/* TPID is global configured value. */
 		case GSW_EXTENDEDVLAN_TREATMENT_VTETYPE:
+			if (!parm->sTreatment.bNewMeterEnable) {
+				switch (parm->sTreatment.eTreatment_4_Tpid_Mode) {
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_1:
+					tbl_prog.val[5] |= (0 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_2:
+					tbl_prog.val[5] |= (1 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_3:
+					tbl_prog.val[5] |= (2 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_4:
+					tbl_prog.val[5] |= (3 << 0x2);
+					break;
+				}
+			}
+
 			switch (parm->sTreatment.sOuterVlan.eDei) {
 			/* DEI (if applicable) is copied from inner VLAN tag of received packet. set 2*/
 			case GSW_EXTENDEDVLAN_TREATMENT_INNER_DEI:
@@ -20306,6 +20372,26 @@ GSW_return_t GSW_ExtendedVlanSet(void *cdev, GSW_EXTENDEDVLAN_config_t *parm)
 
 		/* TPID is global configured value. */
 		case GSW_EXTENDEDVLAN_TREATMENT_VTETYPE:
+			if (!parm->sTreatment.bNewMeterEnable) {
+				switch (parm->sTreatment.eTreatment_4_Tpid_Mode) {
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_1:
+					tbl_prog.val[5] |= (0 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_2:
+					tbl_prog.val[5] |= (1 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_3:
+					tbl_prog.val[5] |= (2 << 0x2);
+					break;
+
+				case GSW_EXTENDEDVLAN_TPID_VTETYPE_4:
+					tbl_prog.val[5] |= (3 << 0x2);
+					break;
+				}
+			}
+
 			switch (parm->sTreatment.sInnerVlan.eDei) {
 			/* DEI (if applicable) is copied from inner VLAN tag of received packet. set 2*/
 			case GSW_EXTENDEDVLAN_TREATMENT_INNER_DEI:
